File: app/src/main/java/party/qwer/Iris/DBObserver.java
package party.qwer.Iris;

import java.io.IOException;
import java.util.concurrent.TimeUnit;
import org.json.JSONException;
import org.json.JSONObject;

public class DBObserver {
    private final KakaoDB kakaoDb;
    private final ObserverHelper observerHelper;
    private Thread pollingThread;

    public DBObserver(KakaoDB kakaoDb, ObserverHelper observerHelper) {
        this.kakaoDb = kakaoDb;
        this.observerHelper = observerHelper;
    }

    public void startPolling() {
        if (pollingThread == null || !pollingThread.isAlive()) {
            pollingThread = new Thread(() -> {
                while (true) {
                    observerHelper.checkChange(kakaoDb);
                    try {
                        long pollingInterval = Configurable.getInstance().getDbPollingRate();
                        Thread.sleep(pollingInterval);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        System.err.println("Polling thread interrupted: " + e.toString());
                        break;
                    }
                }
            });
            pollingThread.setName("DB-Polling-Thread");
            pollingThread.start();
            System.out.println("DB Polling thread started.");
        } else {
            System.out.println("DB Polling thread is already running.");
        }
    }

    public void stopPolling() {
        if (pollingThread != null && pollingThread.isAlive()) {
            pollingThread.interrupt();
            pollingThread = null;
            System.out.println("DB Polling thread stopped.");
        }
    }
}

File: app/src/main/java/party/qwer/iris/DBObserver.kt
package party.qwer.iris

import party.qwer.iris.Configurable.dbPollingRate
import kotlin.concurrent.Volatile

class DBObserver(private val kakaoDb: KakaoDB, private val observerHelper: ObserverHelper) {
    private var pollingThread: Thread? = null

    @Volatile
    private var isObserving: Boolean = false

    fun startPolling() {
        if (pollingThread == null || !pollingThread!!.isAlive) {
            pollingThread = Thread {
                isObserving = true
                while (true) {
                    observerHelper.checkChange(kakaoDb)
                    try {
                        val pollingInterval = dbPollingRate
                        if (pollingInterval > 0) {
                            Thread.sleep(pollingInterval)
                        } else {
                            Thread.sleep(1000)
                        }
                    } catch (e: InterruptedException) {
                        Thread.currentThread().interrupt()
                        System.err.println("Polling thread interrupted: $e")
                        isObserving = false
                        break
                    }
                }
                isObserving = false
            }
            pollingThread!!.name = "DB-Polling-Thread"
            pollingThread!!.start()
            println("DB Polling thread started.")
        } else {
            println("DB Polling thread is already running.")
        }
    }

    fun stopPolling() {
        if (pollingThread != null && pollingThread!!.isAlive) {
            pollingThread!!.interrupt()
            pollingThread = null
            isObserving = false
            println("DB Polling thread stopped.")
        }
    }

    val isPollingThreadAlive: Boolean
        get() = pollingThread != null && pollingThread!!.isAlive
}